presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDifferenceFunction.java

```diff
@@ -0,0 +1,156 @@
+import com.facebook.presto.spi.function.TypeParameter;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import it.unimi.dsi.fastutil.ints.AbstractIntComparator;
+import it.unimi.dsi.fastutil.ints.IntArrays;
+import it.unimi.dsi.fastutil.ints.IntComparator;
+
+@Description("Set difference of two arrays")
+@ScalarFunction("array_diff")
+public final class ArrayDifferenceFunction
+{
+    private static final int INITIAL_SIZE = 128;
+
+    private int[] leftPositions = new int[INITIAL_SIZE];
+    private int[] rightPositions = new int[INITIAL_SIZE];
+    private final PageBuilder pageBuilder;
+
+    @TypeParameter("E")
+    public ArrayDifferenceFunction(@TypeParameter("E") Type elementType)
+    {
+        pageBuilder = new PageBuilder(ImmutableList.of(elementType));
+    }
+
+    private static IntComparator IntBlockCompare(Type type, Block block)
+    {
+        return new AbstractIntComparator()
+        {
+            @Override
+            public int compare(int left, int right)
+            {
+                if (block.isNull(left) && block.isNull(right)) {
+                    return 0;
+                }
+                if (block.isNull(left)) {
+                    return -1;
+                }
+                if (block.isNull(right)) {
+                    return 1;
+                }
+                return type.compareTo(block, left, block, right);
+            }
+        };
+    }
+
+    @TypeParameter("E")
+    @SqlType("array(E)")
+    public Block difference(
+            @TypeParameter("E") Type type,
+            @SqlType("array(E)") Block left,
+            @SqlType("array(E)") Block right)
+    {
+        int leftSize = left.getPositionCount();
+        int rightSize = right.getPositionCount();
+        if (leftSize == 0 || rightSize == 0) {
+            return left;
+        }
+
+        if (leftPositions.length < leftSize) {
+            leftPositions = new int[leftSize];
+        }
+
+        if (rightPositions.length < rightSize) {
+            rightPositions = new int[rightSize];
+        }
+
+        if (pageBuilder.isFull()) {
+            pageBuilder.reset();
+        }
+
+        for (int i = 0; i < leftSize; i++) {
+            leftPositions[i] = i;
+        }
+        for (int i = 0; i < rightSize; i++) {
+            rightPositions[i] = i;
+        }
+        IntArrays.quickSort(leftPositions, 0, leftSize, IntBlockCompare(type, left));
+        IntArrays.quickSort(rightPositions, 0, rightSize, IntBlockCompare(type, right));
+
+        BlockBuilder resultBlockBuilder = pageBuilder.getBlockBuilder(0);
+        int totalCount = 0;
+
+        int leftPositionIndex = 0;
+        while (leftPositionIndex < leftSize && left.isNull(leftPositions[leftPositionIndex])) {
+            resultBlockBuilder.appendNull();
+            leftPositionIndex++;
+            totalCount++;
+        }
+
+        boolean insertNull = right.isNull(rightPositions[0]);
```

Comment:
Let's disallow nulls for the moment, as we discussed.

cc @martint in case Martin disagrees.